src/main/core/controller.rs

```diff
@@ -0,0 +1,492 @@
+                    let _ = &hostname;
+                    let _ = &pcap_dir;
+                }
+
+                for proc in &host.processes {
+                    let plugin_path = CString::new(proc.plugin.to_str().unwrap()).unwrap();
+                    let env = CString::new(&*proc.env).unwrap();
+                    let pause_for_debugging = host.pause_for_debugging;
+
+                    let argv: Vec<CString> = proc
+                        .args
+                        .iter()
+                        .map(|x| CString::new(x.as_bytes()).unwrap())
+                        .collect();
+
+                    // scope used to enforce drop order for pointers
+                    {
+                        let argv_ptrs: Vec<*const i8> = argv
+                            .iter()
+                            .map(|x| x.as_ptr())
+                            // the last element of argv must be NULL
+                            .chain(std::iter::once(std::ptr::null()))
+                            .collect();
+
+                        unsafe {
+                            c::manager_addNewVirtualProcess(
+                                manager,
+                                hostname.as_ptr(),
+                                plugin_path.as_ptr(),
+                                SimulationTime::to_c_simtime(Some(proc.start_time)),
+                                SimulationTime::to_c_simtime(proc.stop_time),
+                                argv_ptrs.as_ptr(),
+                                env.as_ptr(),
+                                pause_for_debugging,
+                            )
+                        }
+
+                        // make sure we never accidentally drop the following objects before running the
+                        // unsafe code (will be a compile-time error if they were dropped)
+                        let _ = &plugin_path;
+                        let _ = &env;
+                        let _ = &argv;
+                        let _ = &argv_ptrs;
+                    }
+                }
+            }
+
+            // we don't use the host/process list anymore, so may as well free the memory
+            hosts.clear();
+            hosts.shrink_to_fit();
+
+            log::info!("Running simulation");
+            unsafe { c::manager_run(manager) };
+            log::info!("Finished simulation, cleaning up now");
+
+            unsafe { c::manager_free(manager) }
+        };
+
+        if rv != 0 {
+            return Err(anyhow::anyhow!("Manager exited with code {}", rv));
+        }
+
+        // access the immutable reference to make sure we haven't mutated self
+        let _fake_ref_for_manager = _fake_ref_for_manager;
+
+        Ok(())
+    }
+}
+
+impl std::ops::Drop for Controller<'_> {
+    fn drop(&mut self) {
+        unsafe { c::dns_free(self.dns) };
+        self.dns = std::ptr::null_mut();
+    }
+}
+
+trait SimController {
```

Comment:
What's the motivation for this trait, vs putting these methods directly in Controller? A comment and/or more-descriptive name might help.